cmd/booster-bitswap/blockfilter/blockfilter.go

```diff
@@ -0,0 +1,244 @@
+		c = cid.NewCidV1(cid.DagProtobuf, c.Hash())
+	}
+	// get base32 string
+	cidStr, err := c.StringOfBase(multibase.Base32)
+	if err != nil {
+		return false, err
+	}
+	// add "/"
+	cidStr += "/"
+	// sha256 sum the bytes
+	shaBytes := sha256.Sum256([]byte(cidStr))
+	// encode to a hex string
+	shaString := hex.EncodeToString(shaBytes[:])
+
+	// check for set inclusion
+	bf.filteredHashesLk.RLock()
+	_, has := bf.filteredHashes[shaString]
+	bf.filteredHashesLk.RUnlock()
+	return has, nil
+}
+
+// fetch deny list fetches and parses a deny list to get a new set of filtered hashes
+// it uses streaming JSON decoding to avoid an intermediate copy of the entire response
+// lenSuggestion is used to avoid a large number of allocations as the list grows
+func (bf *BlockFilter) parseDenyList(denyListStream io.Reader, lenSuggestion int) (map[string]struct{}, error) {
+	// first fetch the reading for the deny list
+	type blockedCid struct {
+		Anchor string `json:"anchor"`
+	}
+	// initialize a json decoder
+	jsonDenyList := json.NewDecoder(denyListStream)
+
+	// read open bracket
+	_, err := jsonDenyList.Token()
+	if err != nil {
+		return nil, fmt.Errorf("parsing denylist: %w", err)
+	}
+
+	filteredHashes := make(map[string]struct{}, lenSuggestion)
+	// while the array contains values
+	for jsonDenyList.More() {
+		var b blockedCid
+		// decode an array value (Message)
+		err = jsonDenyList.Decode(&b)
+		if err != nil {
+			return nil, fmt.Errorf("parsing denylist: %w", err)
+		}
+		// save it in the filtered hash set
+		filteredHashes[b.Anchor] = struct{}{}
+	}
+
+	// read closing bracket
+	_, err = jsonDenyList.Token()
+	if err != nil {
+		return nil, fmt.Errorf("parsing denylist: %w", err)
+	}
+
+	return filteredHashes, nil
+}
+
+// updateDenyList replaces the current filtered hashes after successfully
+// fetching and parsing the latest deny list
+func (bf *BlockFilter) updateDenyList() {
+	fetchTime := time.Now()
+	updated, denyListStream, err := bf.denyListFetcher(bf.lastUpdated)
+	if err != nil {
+		log.Errorf("fetching deny list: %s", err)
+		return
+	}
+	if !updated {
+		return
+	}
+	defer denyListStream.Close()
+	// open the cache file
+	cache, err := os.OpenFile(bf.cacheFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		log.Errorf("opening cache file: %s", err)
+	}
+	defer cache.Close()
+	forkedStream := io.TeeReader(denyListStream, cache)
+	bf.lastUpdated = fetchTime
+	filteredHashes, err := bf.parseDenyList(forkedStream, len(bf.filteredHashes)+expectedListGrowth)
+	if err != nil {
+		log.Errorf("parsing deny list: %s", err)
+		return
+	}
+	bf.filteredHashesLk.Lock()
+	bf.filteredHashes = filteredHashes
+	bf.filteredHashesLk.Unlock()
+}
+
+// run periodically updates the deny list asynchronously
+func (bf *BlockFilter) run(cachedCopy bool) {
+	// if there was a cached copy, immediately asynchornously fetch an update
+	if cachedCopy {
+		bf.updateDenyList()
+	}
+	updater := bf.clock.Timer(UpdateInterval)
```

Comment:
Would it make sense to use [a ticker](https://gobyexample.com/tickers) here instead of a timer? Then you don't need the Reset logic.